src/control/server/ctl_system.go

```diff
@@ -32,64 +32,124 @@ import (
 	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
 )
 
-const stopTimeout = 10 * retryDelay
+const (
+	memberStopTimeout   = 10 * retryDelay
+	prepShutdownTimeout = 10 * retryDelay
+)
+
+// SystemMemberQuery implements the method defined for the Management Service.
+//
+// Return system membership list including member state.
+func (svc *ControlService) SystemMemberQuery(ctx context.Context, req *ctlpb.SystemMemberQueryReq) (*ctlpb.SystemMemberQueryResp, error) {
+	resp := &ctlpb.SystemMemberQueryResp{}
+
+	// verify we are running on a host with the MS leader and therefore will
+	// have membership list.
+	_, err := svc.harness.GetMSLeaderInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	svc.log.Debugf("received SystemMemberQuery RPC; reporting DAOS system members (%+v)", svc.membership.Members())
 
-// systemStop sends multicast KillRank gRPC requests to system membership list.
-func (svc *ControlService) systemStop(ctx context.Context, leader *IOServerInstance) common.SystemMemberResults {
-	members := svc.membership.GetMembers()
+	membersPB, err := common.MembersToPB(svc.membership.Members())
+	if err != nil {
+		return nil, err
+	}
+	resp.Members = membersPB
+
+	svc.log.Debugf("responding to SystemMemberQuery RPC; resp.Members: %+v", resp.Members)
+
+	return resp, nil
+}
+
+// prepShutdown sends multicast PrepShutdown gRPC requests to system membership list.
+func (svc *ControlService) prepShutdown(ctx context.Context, leader *IOServerInstance) common.SystemMemberResults {
+	members := svc.membership.Members()
 	results := make(common.SystemMemberResults, 0, len(members))
 
 	// total retry timeout, allows for 10 retries
-	ctx, _ = context.WithTimeout(ctx, stopTimeout)
+	ctx, _ = context.WithTimeout(ctx, prepShutdownTimeout)
 
-	// TODO: inhibit rebuild on pool services, parallelise and make async.
+	// TODO: parallelise and make async.
 	for _, member := range members {
-		svc.log.Debugf("MgmtSvc.systemStop kill member %+v\n", *member)
+		svc.log.Debugf("MgmtSvc.prepShutdown member %+v\n", *member)
 
-		result := common.SystemMemberResult{ID: member.String(), Action: "stop"}
+		result := common.NewSystemMemberResult(member.String(), "prep shutdown")
 
-		// TODO: force should be applied if a number of retries fail
-		resp, err := leader.msClient.Stop(ctx, member.Addr.String(),
-			&mgmtpb.KillRankReq{Rank: member.Rank, Force: false})
+		resp, err := leader.msClient.PrepShutdown(ctx, member.Addr.String(),
+			&mgmtpb.PrepShutdownReq{Rank: member.Rank})
 		if err != nil {
 			result.Err = err
 		} else if resp.GetStatus() != 0 {
 			result.Err = errors.Errorf("DAOS returned error code: %d\n",
 				resp.GetStatus())
 		}
 
-		if result.Err == nil {
-			svc.membership.Remove(member.Uuid)
-		} else {
-			svc.log.Debugf("MgmtSvc.systemStop error %s\n", result.Err)
+		member.State = common.Stopping
```

Comment:
I think these state values should be named a little more verbosely. `common.Stopping` doesn't read very well. Previously I had suggested moving the membership stuff to a new package, as common is starting to become kind of a grab-bag of random functionality. In any case, at a minimum, I'd recommend prepending the type to the state names, e.g. `common.MemberStateStopping`.

Reply:
I think you'll probably want to make getter/setter methods for SystemMember which protect access to its fields with a mutex. It may not be critical now, but when you're parallelizing things I think that there may be a lot of opportunities for racy access to those fields (State, in particular).